tabs.js

```diff
@@ -0,0 +1,114 @@
+import { getDataBroker } from "./common.js";
+const tabsTemplate = document.createElement('template');
+tabsTemplate.innerHTML = `
+<style>
+:host {
+    --data-color: var(--iobio-data-color, #2d8fc1);
+}
+.iobio-tabs {
+    display: flex;
+    flex-direction: column;
+}
+.tabs {
+    display: flex;
+    justify-content: center;
+    align-items: center;
+    gap: 10px;
+}
+.tab {
+    cursor: pointer;
+    text-align: center;
+    color: grey;
+}
+.tab.active {
+    color:  var(--data-color);
+}
+
+.content-container {
+    position: relative;
+    width: 100%;
+    height: 100%;
+}
+
+::slotted(*) {
+    position: absolute;
+    width: 100%;
+    height: 100%;
+}
+</style>
+    <div class="iobio-tabs">
+        <div class="tabs">
+            <span class="tab tab1-label"></span>  |
+            <span class="tab tab2-label"></span>
+        </div>
+        <div class="content-container">
+            <slot></slot>
+        </div>
+    </div> 
+`;
+
+class Tabs extends HTMLElement {
+    constructor() {
+        super();
+        this.attachShadow({ mode: 'open' });
+        this.shadowRoot.appendChild(tabsTemplate.content.cloneNode(true));
+        
+    }
+  
+    connectedCallback() {
+        this.broker = getDataBroker(this);
+        this.initDOMElements();
+        this.initializeTabs();
+        // show default histogram 
+        this.showChart(0);
+    }
+
+    initDOMElements() {
+        this.tab1_label = this.shadowRoot.querySelector('.tab1-label');
+        this.tab2_label = this.shadowRoot.querySelector('.tab2-label');
+        this.tabs = this.shadowRoot.querySelectorAll('.tab');
+        const slot = this.shadowRoot.querySelector('slot');
+        this.slottedElements = Array.from(slot.assignedElements());
+    }
+
+    initializeTabs() {
+        this.tab1_label.textContent = this.getAttribute('label-1');
+        this.tab2_label.textContent = this.getAttribute('label-2');
+
+        this.tabs.forEach((tab, index) => {
+            tab.addEventListener('click', () => {
+                this.showChart(index);
+                this.updateElementVisibility(index);
+            });
+        });
+    }
+
+    showChart(index) {
+        this.updateTabStyles(index);
+
+        this.broker.onEvent('data-streaming-start', () => {
```

Comment:
Not sure if this was here before and I just missed it, but the iobio-tabs element should not be aware of data brokers. It should be just a dumb container for whatever we want to put inside.

Reply:
yes, it was there before. The tabs elements are aware of the data brokers because all the histogram or percent_box components are set up to be visible after the loading indicator. So when the data streaming starts, the tabs need to be aware of that, then make the unselected tab hidden. Above is my code logic. 

I agree that we need to make the components as decoupling as possible. I'll look into it.

Reply:
Why does the tab container need to know about the state of the contents in order to set the visibility? Don't we just want to always show tab1 and hide tab2 at startup, then only change that if the user clicks?

Reply:
When the data starts steaming, the loading indicator will be "none" and all the svg-containers will be set up to be visible. From there, selected tab will control the visibility of the chart. There is a timing order here.

Reply:
I still don't see the problem. Using an iobio-tabs with 2 histogram children as an example, this is what should happen as I understand it:

1. Startup - histograms both start with loading indicators. However only the first histogram is shown by the tabs, so you see a single loading indicator as intended.
2. Data starts coming in - Both histograms change their internal state to display data. However, the tabs are only showing the first histogram still, so you only see that
3. User selects tab2 - histogram 2 is shown. This will either show the loading indicator or the data, depending on the current state of the histogram. But the tab container doesn't need to know the child internal state. It simply sets the child visibility.

Reply:
Yes, the logic is correct. I get what you're saying. I tried to control the element(here we want any types) visibility only in the slot without changing the child internal state. But the child internal state(e.g, `<div class="iobio-histogram-svg-container" style="visibility: visible;">` ) will still have an effect. So if I don't query by the specific class name, I have no clue how to change the chart's visibility.